### PR TITLE
Add Project Goals RFC

### PR DESCRIPTION
**Pull request by keithamus**
This document outlines the projects goals, and what we hope to do.

[Rendered](https://github.com/kay-framework/rfcs/blob/04d82949362d8a60a00cd38674aaab65699c24b4/0000-project-goals.md)

### CODE REVIEW

File: 0000-project-goals.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
+|  Status  | Proposal Date | Accepted Date | Completed Date | Deprecated Date |
+|:--------:|:-------------:|:-------------:|:--------------:|:---------------:|
+| 'active' |  2014-12-28   |      N/A      |      N/A       |       N/A       |
+
+# Summary
+
+This defines the motivation of the whole Kay-Framework project; it's goals,
```
Comment (JoeChapman) [inline 0000-project-goals.md]: Perhaps consider; 'This document attempts to define the motivation behind the Kay-Framework project...'

and change "it's goals" to 'its goals'

File: 0000-project-goals.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+philosophy and reason for existing.
```
Comment (JoeChapman) [inline 0000-project-goals.md]: You could sound really poncy here if you wanted s/reason for existing/raîson d'etre

File: 0000-project-goals.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+# Motivation
+
+This document should give us a doctrine that define how we operate and what we
```
Comment (JoeChapman) [inline 0000-project-goals.md]: s/define/defines

File: 0000-project-goals.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+do. We can use this as a set of "commandments" to refer to whenever we need to
+make decisions about directions to go in.
```
Comment (JoeChapman) [inline 0000-project-goals.md]: s/about directions to go in/about which direction to follow

File: 0000-project-goals.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+Framework is secure. We want to ensure that users are well protected from any
+potential threats, and are forced into coding their projects to be secure. We
+strive to fix security issues promptly to protect users and maintain their
+trust.
+
+## Stable
+
+All code should maintain a high level of stability - ensuring that any breaking
+features are correctly documented as such. All of our software and all
+dependencies should adhere to SemVer versioning to ensure our users' projects do
```
Comment (JoeChapman) [inline 0000-project-goals.md]: s/SemVer versioning/semVer. It's like saying 'semantic versioning versioning. Isn't it unnecessary?

File: 0000-project-goals.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+not break.
+
+Any unexpected errors from our software should be dealt with promptly.
+Applications built with Kay Framework should be crash and leak free.
```
Comment (JoeChapman) [inline 0000-project-goals.md]: Should all instances of 'Kay Framework' be hyphenated to 'Kay-Framework'? If so, there are some instances of inconsistency

File: 0000-project-goals.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+## Safe
+
+Kay Framework should encourage users to code in a safe way. This means users
+should be prevented from coding dangerous edge cases or attack vectors in their
+code. The path of least resistence, through code and throughout documentation
+should strive to keep the user from shooting themselves in the foot.
```
Comment (JoeChapman) [inline 0000-project-goals.md]: s/themselves/themself

File: 0000-project-goals.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+## Performant
+
+Speed and memory use is are an ever important goal for Kay Framework.
```
Comment (JoeChapman) [inline 0000-project-goals.md]: unnecessary use of 'are'

File: 0000-project-goals.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+Performance can never be "good enough", but always be improved. Our software
+should be architected to avoid performance issues as applications become more
+complex, and strive for the fastest code for both small and large applications.
+
+## Open Source
+
+Kay Framework should remain free for anyone to use, both for personal and
+proprietary projects and applications. All of Kay Framework is MIT licensed. All
+of Kay Frameworks dependencies are either MIT, BSD (2/3 clause) or ISC licensed.
```
Comment (JoeChapman) [inline 0000-project-goals.md]: Plural? 'Kay Frameworks'

File: 0000-project-goals.md [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+## Open Governance
+
+Kay Framework should have an Open Governance model. All work, including work in
+progress should be freely available to collaborate on by the entire community.
+Any new features developed by Kay Framework should be open for discussion and
+improvement by the entire community. Any new releases should be handled by the
+community.
+
+Core maintainers hold responsibility to guide the community, but carry no power
+over it. They are servants, not masters. This means core maintainers will never
```
Comment (JoeChapman) [inline 0000-project-goals.md]: Sounds very dramatic 'They are servants, not masters'. IMO, this should be refactored out.

File: 0000-project-goals.md (hunk 7 above)
Comment (keithamus) [inline 0000-project-goals.md]: Not sure about this one. Seems as though "shooting themselves in the foot" is correct

**Comment by JoeChapman** [comment]
A succinct, description of the project overview, however, I think it lacks the detail required to convey the significance of the objectives. As vaguely eluded to in comments such as; "Sounds very dramatic 'They are servants, not masters'. IMO, this should be refactored out.", there is a very fine line between conveying passion and technical objectivity in prose. Be voracious, but refined when depicting the project goals in light of the "community" element. We need to appear accepting and cooperative but not desperate for collaboration. I believe people/contributors will want a clear direction so we shouldn't leave too much open for interpretation.

**Comment by keithamus** [comment]
RFCs should take a layered approach, and so the detail you say is lacking may just be out of scope of this layer. The idea behind this RFC is to set a baseline of things we should strive as a team and a methodology for coding & working together. As the whole project evolves RFCs will get more concrete. 

I guess you could imagine this as focussing a light - the first few RFCs will be broad focus, almost "ambient", but the more we do the narrower the beam gets. By the time we've got a few dozen/hundred RFCs they'll be talking about specific features, complete with syntax proposals. New contributors can read through the RFCs to gain context, or just dive right into the latest RFC and start implementing.

**Comment by JoeChapman** [comment]
For context and posterity, following is the comment I made in Skype that I think your previous comment makes reference to, and which, I think is an appropriate response. However, I don't believe it addresses my comment on 'passion' & 'objectivity':

[08/01/2015 09:47:50] Joseph Philip Chapman: On 08/01/2015, at 09:42, Keith Cirkel wrote:
> But ultimately its to create a web framework, like Django/Rails/Symfony/CakePHP
> A Full Stack web framework
> so Models, Controllers, Views, Forms, Validation, Emailing, Templating, etc etc

sorry, I meant this too ^
[08/01/2015 09:49:02] Joseph Philip Chapman: I understand you want many details it to be defined by committee, but I want to see a bit more technical direction in the doc.

___________________________________________________________________________________

**Comment by keithamus** [comment]
I'm working on one now which goes into a little more detail about what building blocks we need for a good framework, which will narrow focus of these goals (or philosophy) down into more solid objectives.

**Pull request merged**